clients/cli/src/main/java/org/apache/gravitino/cli/outputs/TableFormat.java

```diff
@@ -18,203 +18,581 @@
+    } else if (entity instanceof Catalog[]) {
+      new CatalogListTableFormat(context).output((Catalog[]) entity);
     } else {
       throw new IllegalArgumentException("Unsupported object type");
     }
   }
 
-  static final class MetalakeTableFormat implements OutputFormat<Metalake> {
-    @Override
-    public void output(Metalake metalake) {
-      List<String> headers = Arrays.asList("metalake", "comment");
-      List<List<String>> rows = new ArrayList<>();
-      rows.add(Arrays.asList(metalake.name(), metalake.comment()));
-      TableFormatImpl tableFormat = new TableFormatImpl();
-      tableFormat.print(headers, rows);
+  /**
+   * Creates a new {@link TableFormat} with the specified properties.
+   *
+   * @param context the command context.
+   */
+  public TableFormat(CommandContext context) {
+    super(context);
+    // TODO: add other options for TableFormat
+  }
+
+  /**
+   * Get the formatted output string for the given columns.
+   *
+   * @param columns the columns to print.
+   * @return the table formatted output string.
+   */
+  public String getTableFormat(Column... columns) {
+    checkColumns(columns);
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+    String[] headers =
+        Arrays.stream(columns)
+            .map(Column::getHeader)
+            .filter(Objects::nonNull)
+            .toArray(String[]::new);
+
+    List<Character> borders = OutputConstant.BASIC_ASCII;
+    checkHeaders(headers, columns);
+
+    if (headers.length != columns.length) {
+      throw new IllegalArgumentException("Headers must be provided for all columns");
+    }
+
+    if (limit != -1) {
+      columns = getLimitedColumns(columns);
     }
+
+    try (OutputStreamWriter osw = new OutputStreamWriter(baos, StandardCharsets.UTF_8)) {
+      writeUpperBorder(osw, borders, System.lineSeparator(), columns);
+      writeHeader(osw, borders, System.lineSeparator(), columns);
+      writeHeaderBorder(osw, borders, System.lineSeparator(), columns);
+      writeData(osw, borders, columns, System.lineSeparator());
+      writeBottomBorder(osw, borders, System.lineSeparator(), columns);
+
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return new String(baos.toByteArray(), StandardCharsets.UTF_8);
   }
 
-  static final class MetalakesTableFormat implements OutputFormat<Metalake[]> {
-    @Override
-    public void output(Metalake[] metalakes) {
-      if (metalakes.length == 0) {
-        System.out.println("No metalakes exist.");
-      } else {
-        List<String> headers = Collections.singletonList("metalake");
-        List<List<String>> rows = new ArrayList<>();
-        for (int i = 0; i < metalakes.length; i++) {
-          rows.add(Arrays.asList(metalakes[i].name()));
-        }
-        TableFormatImpl tableFormat = new TableFormatImpl();
-        tableFormat.print(headers, rows);
-      }
+  private void checkColumns(Column... columns) {
```

Comment:
I'm not sure that the Preconditions calls are adding anything. What does the user see if one of these calls fails?